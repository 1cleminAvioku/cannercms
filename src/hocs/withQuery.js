// @flow

import * as React from 'react';
import PropTypes from 'prop-types';
import {Spin, Icon} from 'antd';
const antIcon = <Icon type="loading" style={{fontSize: 24}} spin />;

type Props = {
  id: string,
  isEntity: boolean,
  name: string,
  title: string, // only used in withTitleAndDesription hoc
  description: string, // only used in withTitleAndDesription hoc
  layout?: 'horizontal' | 'vertical', // only used in withTitleAndDesription hoc
  [string]: any
};

type State = {
  value: any,
  rootValue: any,
  isFetching: boolean
}

type Context = {
  fetch: (key: string, componentId: string, query: any) => Promise<any>,
  subscribe: (key: string, componentId: string, type: string, callback: Function) => Promise<any>,
  query: {
    filter?: any,
    sort?: any,
    pagination?: any
  },
  componentId: string
}

export default function withQuery(Com: React.ComponentType<*>) {
  // this hoc will fetch data;
  return class ComponentWithQuery extends React.PureComponent<Props, State> {
    componentId: string;
    key: string;
    subscription: any;
    static contextTypes = {
      fetch: PropTypes.func,
      subscribe: PropTypes.func,
      query: PropTypes.shape({
        filter: PropTypes.any,
        sort: PropTypes.any,
        pagination: PropTypes.any
      }),
      componentId: PropTypes.string
    };

    static childContextTypes = {
      componentId: PropTypes.string
    }

    getChildContext() {
      return {
        componentId: this.context.componentId || this.props.id
      };
    }

    constructor(props: Props, context: Context) {
      super(props);
      // now use is to be a subjectId
      this.state = {
        value: null,
        rootValue: null,
        isFetching: true
      };
      this.componentId = context.componentId || props.id;
      this.key = props.id.split('/')[0];
    }

    componentDidMount() {
      this.queryData();
    }

    componentWillUnmount() {
      this.unsubscribe();
    }

    queryData = () => {
      const {fetch, query} = this.context;
      // second key: use key as a subjectID
      fetch(this.key, this.componentId, query).then(ctx => {
        this.setState({
          rootValue: ctx.response.body,
          value: getValue(ctx.response.body, this.props.id),
          isFetching: false
        });
        this.subscribe();
      });
    }

    unsubscribe = () => {
      if (this.subscription) {
        this.subscription.unsubscribe();
      }
    }

    subscribe = (): Promise<any> => {
      const {subscribe} = this.context;
      return subscribe(this.key, this.componentId, 'value', value => {
        this.setState({
          rootValue: value,
          value: getValue(value, this.props.id)
        });
      }).then(subscription => {
        this.subscription = subscription;
      });
    }

    render() {
      const {value, isFetching, rootValue} = this.state;
      if (isFetching) {
        return <Spin indicator={antIcon} />;
      }
      return <Com {...this.props} rootValue={rootValue} value={value} />;
    }
  };
}

function getValue(value, id) {
  if (value && id) {
    return value.getIn(id.split('/').slice(1));
  }
  return null;
}