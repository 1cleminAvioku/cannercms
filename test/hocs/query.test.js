import * as React from 'react';
import Enzyme, { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import Adapter from 'enzyme-adapter-react-16';
import withQuery from '../../src/hocs/query';
import {fromJS} from 'immutable';
import RefId from 'canner-ref-id';

Enzyme.configure({ adapter: new Adapter() });

describe('with  query', () => {
  let WrapperComponent, props, MockComponent, mockFetch,
    mockUnsubscribe, mockSubscribe;

  beforeEach(() => {
    MockComponent = function MockComponent() {
      return (<div>Component</div>);
    }
    mockFetch = jest.fn().mockImplementation(() => Promise.resolve(fromJS({
      posts: [{
        id: 'post1',
        title: 'POST1'
      }]
    })));
    mockUnsubscribe = jest.fn();
    mockSubscribe = jest.fn().mockImplementation(() => ({
      unsubscribe: mockUnsubscribe
    }))
    const rootValue = fromJS({
      posts: [{
        id: 'id1',
        title: 'post1'
      }]
    });
    props = {
      refId: new RefId('posts'),
      value: rootValue,
      rootValue,
      relation: undefined,
      items: {
        title: {
          tyep: 'string'
        }
      },
      fetch: mockFetch,
      subscribe: mockSubscribe,
      unsubscribe: mockUnsubscribe
    }
    WrapperComponent = withQuery(MockComponent);
  });

  it('should render', () => {
    const wrapper = shallow(<WrapperComponent {...props} />);
    expect(toJson(wrapper)).toMatchSnapshot();
    wrapper.setState({
      isFetching: false
    });
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('queryData when mount', () => {
    shallow(<WrapperComponent {...props} />);
    expect(mockFetch).toHaveBeenCalledWith('posts');
  });

  it('subscribe should work this.subscribtion exists', () => {
    const wrapper = shallow(<WrapperComponent {...props} />);
    wrapper.instance().subscribe();
    expect(mockSubscribe).toHaveBeenCalledWith('posts', expect.anything(Function));
    expect(wrapper.instance().subscription).toHaveProperty('unsubscribe');
  });

  it('call unscribe should works', () => {
    const wrapper = shallow(<WrapperComponent {...props} />);
    wrapper.instance().subscribe();
    wrapper.instance().unsubscribe();
    expect(mockUnsubscribe).toHaveBeenCalledTimes(1);
  });
});
