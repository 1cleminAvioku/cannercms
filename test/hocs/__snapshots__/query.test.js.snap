// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`with  query should render 1`] = `
<Spin
  indicator={
    <Icon
      spin={true}
      style={
        Object {
          "fontSize": 24,
        }
      }
      type="loading"
    />
  }
  prefixCls="ant-spin"
  size="default"
  spinning={true}
  wrapperClassName=""
/>
`;

exports[`with  query should render 2`] = `
<MockComponent
  fetch={
    [MockFunction] {
      "calls": Array [
        Array [
          "posts",
        ],
      ],
    }
  }
  items={
    Object {
      "title": Object {
        "tyep": "string",
      },
    }
  }
  query={
    Query {
      "getQueries": [Function],
      "queries": Object {
        "posts": Object {
          "alias": "posts",
          "args": Object {
            "pagination": Object {
              "first": 10,
            },
          },
          "connection": true,
          "fields": Object {
            "id": null,
            "title": null,
          },
          "isPlural": true,
        },
      },
      "schema": Object {
        "posts": Object {
          "items": Object {
            "items": Object {
              "title": Object {
                "type": "string",
              },
            },
            "type": "object",
          },
          "type": "array",
        },
      },
      "toGQL": [Function],
      "updateQueries": [Function],
    }
  }
  refId={
    CannerRefId {
      "_internalArr": Array [
        "posts",
      ],
      "child": [Function],
      "getPathArr": [Function],
      "remove": [Function],
      "toString": [Function],
    }
  }
  rootValue={null}
  subscribe={[MockFunction]}
  unsubscribe={[MockFunction]}
  value={null}
/>
`;
